Marlin/Cartridge.cpp

```diff
@@ -0,0 +1,237 @@
+        cartridgePresentUpdate(0);
+    } else {
+        cartridgeAbsentUpdate(0);
+    }
+    // Cartridge one is pulled high by default
+    if (READ(CART1_SIG2_PIN) == LOW) {
+        cartridgePresentUpdate(1);
+    } else {
+        cartridgeAbsentUpdate(1);
+    }
+}
+
+/**
+ * This function checks to see if a cartridge has been removed from the
+ * system, allowing us to make judgement calls for error reporting. This
+ * information is updated by calling UpdateCartridgeStatus(). Will also 
+ * return true if no cartridges are present. This version includes hysteresis
+ * so that when the cartridge is re-inserted, we don't immediately throw a 
+ * temperature error (we'll read max temp when a cartridge is removed and
+ * this may not update immediately when we call this function)
+ * @returns    Returns true if a cartridge has been removed, or no 
+ *             cartridge is present WITH HYSTERESIS
+ */
+bool CartridgeRemoved(void) {
+    bool returnValue = false;
+    static unsigned int cartridgeRemovalHysteresis = 0;
+    bool removedCondition = (!cartridgePresent[0] || !cartridgesPresentCheck());
+
+    // If a cartridge is seen to be removed, set the hysteresis counter.
+    // If the appropriate cartridges are absent but we don't see that a
+    // cartridge has been removed, we started up with it missing and will also
+    // mark it as removed.
+    if (cartridgesRemovedCheck()) {
+        cartridgeRemovalHysteresis = CARTRIDGE_REMOVAL_HYSTERESIS_COUNT;
+    } else if (removedCondition) {
+        cartridgeRemovalHysteresis = CARTRIDGE_REMOVAL_HYSTERESIS_COUNT;
+    }
+
+    // We have this hysteresis here to accomodate putting the cartridge
+    // back in. Without it, Marlin will recognize the cartridge has been
+    // reinserted before the temperature updates from its maximum value and
+    // will throw a high temperature error.
+    if (cartridgeRemovalHysteresis > 0) {
+        cartridgeRemovalHysteresis--;
+        returnValue = true;
+    }
+
+    return returnValue;
+}
+
+/**
+ * This function checks to see if the FFF cartridge is removed,
+ * to prevent heating
+ * @returns    Returns true if an FFF cartridge has been removed
+ */
+bool CartridgeRemovedFFF(void) {
+    if (!cartridgePresent[0]) {
+        return true;
+    }
+    return false;
+}
+
+/**
+ * This is the error handler for when we see the cartridge removed error. It 
+ * reports to octoprint that we should pause the print, as well as disabling
+ * all heaters and stopping the currently queued commands. This will only get
+ * called once if it's been called multiple times in quick succession.
+ * @inputs     An input that will be displayed on the serial monitor
+ */
+void _cartridge_removed_error(const char *serial_msg) {
+    if (IsSafetyCriticalSection()) {
+        static bool killed = false;
+        if (IsRunning()) {
+            SERIAL_ERROR_START;
+            serialprintPGM(serial_msg);
+            SERIAL_EOL;
+        }
+        if (!killed) {
```

Comment:
@jminardi in C and C++, you can use the `static` keyword when you declare a local variable to allocate a memory location for that variable (instead of just using the program stack), allowing persistence. So the line `static bool killed = false;` initializes `killed` to `false`, and then never runs that initialization again.